File: 10_generics/dotnet/tests/ExampleTests.cs
using NUnit.Framework;
using tests;

namespace Tests
{
    public class Tests
    {
        [Test]
        public void testTuple() // this test passes
        {
            MyTuple<int> intTuple1 = new MyTuple<int>(1, 2);
            MyTuple<int> intTuple2 = new MyTuple<int>(1, 2);
            MyTuple<string> stringTuple = new MyTuple<string>("first", "second");

            Assert.AreNotEqual(intTuple1.GetType(), stringTuple.GetType());
            Assert.AreEqual(intTuple1.GetType(), intTuple2.GetType());
        }

        [Test]
        public void testStaticFields()
        {
            MyTuple<int> intTuple = new MyTuple<int>(1, 2);
            MyTuple<int>.staticVal = "int tuple val";
            MyTuple<string>.staticVal = "string tuple val";

            Assert.AreNotEqual(MyTuple<int>.staticVal, MyTuple<string>.staticVal);
        }
    }
}

File: 10_generics/javatests/src/test/java/javatests/ExampleTests.java
package javatests;

import org.junit.Test;

import static org.junit.Assert.assertEquals;

public class ExampleTests {

    @Test
    public void testTypeErasure() {
        MyTuple<Integer> intTuple = new MyTuple<>(1,2);
        MyTuple<String> stringTuple = new MyTuple<>("first","second");

        assertEquals(intTuple.getClass(), stringTuple.getClass()); // hmmmm...
    }

    @Test
    public void staticFields() {
        MyTuple<Integer> intTuple = new MyTuple<>(1,2);
        intTuple.globalStaticField = "int tuple set";
        MyTuple<String> stringTuple = new MyTuple<>("first", "second");
        stringTuple.globalStaticField = "string tuple set";

        assertEquals("string tuple set", intTuple.globalStaticField);
    }

}
